### PR TITLE
Adding Restrict access to Amazon EC2 metadata piece

### PR DESCRIPTION
**Pull request by karensawrey**
Attempt #1.

Please take a good look at this one, @pda! I've added the changes that you and @chloeruka recommended. Thanks!

### CODE REVIEW

**Review by plaindocs — DISMISSED** [review]
There is a ton of useful information packed into this pull request, but I think we need to digest it a little further. 

Let's highlight the problem more directly **restrict access to your Amazon Elastic Compute Cloud (EC2) metadata**. Those intro paragraphs are super dense, and might need a little unpacking.

Then when presenting the two options, we need to make them not look sequential (because they are either / or) and be more clear about when you'd use them. The second is *not Elastic Stack for AWS*, does it follow that the first is for *Elastic Stack for AWS* ?

We don't use the term spinning up instances or agents elsewhere, let's find out what we do say and be consistent through out. Probably 'starting' works. 

Other questions

* What is " CI stack approach"
* How would you know what size to make the pool

This is a great start on some docs for a complex topic.

File: pages/agent/v3/aws.md.erb [hunk 1]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+
+If you run all the build steps in Docker containers, you should use the latter approach for the best results. And if you are using Kubernetes for your Buildkite CI, you can check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.  
+
+### Restricting permissions via compartmentalization of agents
+
+This approach suggests the use of Elastic Stack. However, these instructions can also be followed using hooks or scripts.
+
+You can divide your Buildkite agents by responsibilities. For example — agents building for development environments or release, agents deploying for staging or production, etc. This will help reflect multiple AWS environments at your Buildkite organization. 
+
+To divide the responsibilities and permissions of Buildkite agents and provide the relevant teams with sandboxed IAM permissions for their own microservices, for each pipeline you will need to use a [third-party AWS AssumeRole Buildkite Plugin] (https://github.com/cultureamp/aws-assume-role-buildkite-plugin/). This plugin also takes care of the injection of AWS credentials.
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:44]:
```suggestion
To divide the responsibilities and permissions of Buildkite agents and provide the relevant teams with sandboxed IAM permissions for their own microservices, for each pipeline you will need to use a [third-party AWS AssumeRole Buildkite Plugin](https://github.com/cultureamp/aws-assume-role-buildkite-plugin/). This plugin also takes care of the injection of AWS credentials.   
```

File: pages/agent/v3/aws.md.erb [hunk 2]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+
+### Restricting permissions via compartmentalization of agents
+
+This approach suggests the use of Elastic Stack. However, these instructions can also be followed using hooks or scripts.
+
+You can divide your Buildkite agents by responsibilities. For example — agents building for development environments or release, agents deploying for staging or production, etc. This will help reflect multiple AWS environments at your Buildkite organization. 
+
+To divide the responsibilities and permissions of Buildkite agents and provide the relevant teams with sandboxed IAM permissions for their own microservices, for each pipeline you will need to use a [third-party AWS AssumeRole Buildkite Plugin] (https://github.com/cultureamp/aws-assume-role-buildkite-plugin/). This plugin also takes care of the injection of AWS credentials.   
+
+To ensure that the agent in charge of a job, build, pipeline, etc., is allowed to run and will assume the role it has permission to, you can perform a pre-checkout [hook on the agent](https://buildkite.com/docs/agent/v3/hooks).
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:46]:
```suggestion
To ensure that the agent in charge of a job, build, pipeline, etc., is allowed to run and will assume the role it has permission to, you can perform a [pre-checkout hook](https://buildkite.com/docs/agent/v3/hooks) on the agent.
```

File: pages/agent/v3/aws.md.erb [hunk 3]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+
+### Restricting permissions by downgrading an instance profile role  
+
+This approach [is suggested by Amazon](https://docs.aws.amazon.com/cli/latest/reference/ec2/replace-iam-instance-profile-association.html) and will be helpful if you are not using Elastic Stack for AWS.
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:50]:
```suggestion
This approach is [suggested by Amazon](https://docs.aws.amazon.com/cli/latest/reference/ec2/replace-iam-instance-profile-association.html) and is helpful if you are not using Elastic CI Stack for AWS. 
```

File: pages/agent/v3/aws.md.erb [hunk 4]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+
+To ensure that the agent in charge of a job, build, pipeline, etc., is allowed to run and will assume the role it has permission to, you can perform a pre-checkout [hook on the agent](https://buildkite.com/docs/agent/v3/hooks).
+
+### Restricting permissions by downgrading an instance profile role  
+
+This approach [is suggested by Amazon](https://docs.aws.amazon.com/cli/latest/reference/ec2/replace-iam-instance-profile-association.html) and will be helpful if you are not using Elastic Stack for AWS. 
+
+To restrict the permissions, you can permanently downgrade an instance’s profile from a high-permission bootstrap role to a low-permission steady-state role. The high-permission role has a policy that allows replacing the instance profile with a low-permission role, but there is no such policy for a low-permission role.
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:52]:
```suggestion
To restrict permissions of an instance, you can permanently downgrade an instance’s profile from a high-permission bootstrap role to a low-permission steady-state role. The high-permission role has a policy that allows replacing the instance profile with a low-permission role, but there is no such policy for a low-permission role.
```

File: pages/agent/v3/aws.md.erb [hunk 5]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+
+## Further tightening the security around EC2 permissions  
+
+For even better security, you can expire agents after a job. For example, you can:
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:56]:
```suggestion
For added security, you can expire agents after a job. For example, you can:  
```

**Review by plaindocs — COMMENTED** [review]
This is great. I've made some minor suggestions that we can tweak, but but otherwise we're almost good to go!

File: pages/agent/v3/aws.md.erb [hunk 6]
```diff
@@ -21,3 +21,56 @@ button on your organization’s Agents page.
...
+For security reasons, you may choose to restrict access to Amazon EC2 metadata that your builds have if you provision various elements of your infrastructure (RDS, Redis, SQS, etc.) using AWS permission sandboxes (a role with many permissions to EC2 instances).
+
+If you run your builds on an AWS EC2 permission sandbox and then allow Buildkite agents to generate and inject some sandboxed AWS credentials into the build secrets, such builds will have access to the EC2 metadata API. They will also be able to access the same permissions as your EC2. 
+
+To avoid this, you need to prevent the builds from accessing your EC2 metadata or provide sandboxed AWS credentials for each build and restrict their permissions. There are two main ways to do it:
+
+* Via compartmentalization of Buildkite agents
+* Through downgrading an instance profile role
+
+If you run all the build steps in Docker containers, you should use the latter approach for the best results. And if you are using Kubernetes for your Buildkite CI, you can check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.
```
[the remark this answers is not in the document]
Comment (karensawrey) [reply pages/agent/v3/aws.md.erb:36]: The addition about Kubernetes is the 2nd approach, I'll rewrite this bit for improved clarity,

**Review by plaindocs — APPROVED** [review]
LGTM

File: pages/agent/v3/aws.md.erb [hunk 7]
```diff
@@ -15,3 +15,63 @@ The [Elastic CI Stack for AWS](/docs/agent/v3/elastic_ci_aws) is a pre-built Clo
...
+
+If you run your builds on an AWS EC2 permission sandbox and then allow Buildkite agents to generate and inject some sandboxed AWS credentials into the build secrets, such builds will have access to the EC2 metadata API. They will also be able to access the same permissions as your EC2. 
+
+To avoid this, you need to prevent the builds from accessing your EC2 metadata or provide sandboxed AWS credentials for each build and restrict their permissions. There are two main ways to do it:
+
+* Compartmentalizing your Buildkite agents
+* Downgrading an instance profile role
+
+If you run all the build steps in Docker containers, take a look at [compartmentalizing your agents](/docs/agent/v3/aws#restricting-permissions-via-compartmentalization-of-agents). If you are using Kubernetes for your Buildkite CI, use the [same approach](/docs/agent/v3/aws#restricting-permissions-via-compartmentalization-of-agents) and also check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:36]:
```suggestion
If you run all the build steps in Docker containers, take a look at [compartmentalizing your agents](#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents). If you are using Kubernetes for your Buildkite CI, use the [same approach](/#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents) and also check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.  
```

File: pages/agent/v3/aws.md.erb [hunk 8]
```diff
@@ -15,3 +15,62 @@ The [Elastic CI Stack for AWS](/docs/agent/v3/elastic_ci_aws) is a pre-built Clo
...
+
+If you run your builds on an AWS EC2 permission sandbox and then allow Buildkite agents to generate and inject some sandboxed AWS credentials into the build secrets, such builds will have access to the EC2 metadata API. They will also be able to access the same permissions as your EC2. 
+
+To avoid this, you need to prevent the builds from accessing your EC2 metadata or provide sandboxed AWS credentials for each build and restrict their permissions. There are two main ways to do it:
+
+* Compartmentalizing your Buildkite agents
+* Downgrading an instance profile role
+
+If you run all the build steps in Docker containers, take a look at [compartmentalizing your agents](#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents). If you are using Kubernetes for your Buildkite CI, use the [same approach](/#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents) and also check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.
```
Comment (plaindocs) [inline pages/agent/v3/aws.md.erb:36]:
```suggestion
If you run all the build steps in Docker containers, take a look at [compartmentalizing your agents](#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents). If you are using Kubernetes for your Buildkite CI, use the [same approach](#preventing-builds-from-accessing-amazon-ec2-metadata-restricting-permissions-via-compartmentalization-of-agents) and also check out [this article](https://github.com/blakestoddard/scaledkite) for more information and inspiration.  
```

**Pull request merged**